### DISCUSSION TITLE
Trying to debug Timeout when calling IPeripheral.GetKnownCharacteristic

### DISCUSSION THREAD

**Discussion by munkii** [shinyorg/shiny#1190] [Q&A]
I've spent a day or so trying to get to the bottom of an Android Bluetooth issue that will not replicate when the debugger is attached and I was hoping someone here might be able to suggest some areas to check.

The intial "connection" to a bluetooth peripheral always works. It is done from a ViewModel and when the View/ViewModel closes I can see the Notify being stopped, disposed and the Peripheral being disconnected.

On any subsequent loads of that ViewModel, attempts to connect to the Peripheral succeed but the call to GetKnownCharacteristic fails. This is how we are getting the Characteristic

```
var nativeCharacteristic =
                        await this.device
                                    .GetKnownCharacteristic(serviceUuid.ToString(), characteristicUuid.ToString(), true)
                                    .Take(1)
                                    .Timeout(TimeSpan.FromSeconds(10));
```

Looking at the Android Device Log I can see the following Error when the call times out

> 02-07 12:58:19.963	Samsung SM-G975F	Error	7379	bt_stack	[ERROR:gatt_api.cc(1237)] GATT_Connectgatt_if=12, address=00:80:25:d5:4a:19, is_direct=1, opportunistic=0

When the connection works OK I see these errors in the Device Log

> 02-07 12:59:49.218	Samsung SM-G975F	Error	7379	bt_stack	[ERROR:bta_gattc_cache.cc(974)] bta_gattc_cache_load: can't open GATT cache file /data/misc/bluetooth/gatt_cache_008025d54a19 for reading, error: No such file or directory
> 02-07 12:59:46.845	Samsung SM-G975F	Error	7379	bt_stack	[ERROR:gatt_api.cc(1237)] GATT_Connectgatt_if=11, address=00:80:25:d5:4a:19, is_direct=1, opportunistic=0
> 

Other things I know.

- The device itself is not getting into an invalid state because restarting the app without power cycling the device gets us back to a working situation
- The exception thrown is System.TimeoutException from System.Reactive.Subjects.AsyncSubject

I appreciate this is not much to go on but any pointers would be helpful.

**Comment by aritchie** [comment]
I have not seen this specific error with Android before, but it looks like a traditional race condition that exists within the android stack.  

1. After disconnecting, are you triggering a scan operation anywhere?  Anytime you trigger a scan, you should get rid of any tracked instances you have to a device.
2. What version of Android/Device is this?

**Reply by munkii** [reply]
I'm not triggering a scan and we see it on Android 9, 12 and 13 on Pixel and Samsung.

>  traditional race condition that exists within the android stack.

Who doesn't like trying to debug a race condition :-(

**Reply by aritchie** [reply]
Can you include a reproducible sample and/or all of the android bluetooth logs right up to these errors?

**Reply by munkii** [reply]
I wish I could but the code is way too tightly coupled. I inherited this part of the codebase (I would say that woudn't I) and it already had lines like this after certain BT calls.

```
// HACK Delay added because android bluetooth needs time between calls
await Task.Delay(750);
```

In fairness it was working but I stupidly decided to "shake the jello" and move from the old PlugIn.BluetoothLe library to the later Shiny.BluetoohtLe. I should have left well alone :-)

All we need to do is connect to a couple of medical devices, listen for data notifcations and disconnect. This connecting,listening and disconnecting is done a couple of times a day. If I was to start from scratch would you suggest using the Managed classes and starting with the BluetoothLE-Managed sample?

I did try that sample yesterday but couldn't find the EnableNotifications method on iPeripheral. 

I realise this many questions probably does stray away from what these discussions are for and too close to consultancy/suppport.  Just ignore the bits where i've pushed it too far

**Reply by aritchie** [reply]
So the only real difference in terms of implementation that would cause this has to do with settings in here:

https://github.com/shinyorg/shiny/blob/v2.7.0/src/Shiny.BluetoothLE/BleConfiguration.cs

Ensure that you are using the internal sync queue and possibly main thread is set to true.  If you are doing a lot of serial operations, it will slow them down quite dramatically.  If you are using BLE "the proper way" for tiny sets of data, you won't be bothered at all by these settings.

The managed scanner is great for all scenarios, the managed peripheral is meant for long workloads with a peripheral.  I would stick to standard if that's what your code base already does.

**Reply by aritchie** [reply]
> I realise this many questions probably does stray away from what these discussions are for and too close to consultancy/suppport. Just ignore the bits where i've pushed it too far

You're a long time sponsor which I appreciate.  If the problem is too big, I will ask for breakdowns where I can.  I'm willing to help sponsors where I can by giving some suggestion, but you can only go so far without a repro anyhow.

**Reply by munkii** [reply]
We've made some progress and we have a repro sample. If on closing our ViewModel we don't call CancelConnection and only dispose the WhenStatusChanged and Notify observables subsequent calls to GetKnownCharacteristic do not timeout. However they don't result in the Notify call actually notifying either. 

From the Debug Logs.

When the GetKnownCharacteristic and Notify subscription work the logs look like this

> [0:] About to try connecting to Charcteristic
> [BluetoothGatt] discoverServices() - device: 00:80:25:D5:4A:19
> [BluetoothGatt] onConnectionUpdated() - Device=00:80:25:D5:4A:19 interval=6 latency=0 timeout=500 status=0
> [BluetoothGatt] onSearchComplete() = Device=00:80:25:D5:4A:19 Status=0
> [BluetoothGatt] onConnectionUpdated() - Device=00:80:25:D5:4A:19 interval=24 latency=0 timeout=16 status=0
> [0:] Got nativeCharacteristic 
> [BluetoothGatt] setCharacteristicNotification() - uuid: 00000002-0000-1000-8000-008025000000 enable: true
> [0:] nativeCharacteristic Notify enabled
> 

When the GetKnownCharacteristic is going to Timeout the logs looks like this

> [0:] About to try connecting to Charcteristic
> [BluetoothGatt] onConnectionUpdated() - Device=00:80:25:D5:4A:19 interval=12 latency=0 timeout=500 status=0
> [BluetoothGatt] onConnectionUpdated() - Device=00:80:25:D5:4A:19 interval=6 latency=0 timeout=500 status=0
> [BluetoothGatt] onConnectionUpdated() - Device=00:80:25:D5:4A:19 interval=24 latency=0 timeout=16 status=0
> 
> **System.TimeoutException:** 'Loading...'
> 

If we do not disconnect and on subsequent load of the ViewModel we simply call GetKnownCharacteristic and Notify on the connected KnownPeripheral the logs look like this 

> [0:] About to try connecting to Charcteristic
> [BluetoothGatt] discoverServices() - device: 00:80:25:D5:4A:19
> [BluetoothGatt] onSearchComplete() = Device=00:80:25:D5:4A:19 Status=0
> [0:] Got nativeCharacteristic 
> [BluetoothGatt] setCharacteristicNotification() - uuid: 00000002-0000-1000-8000-008025000000 enable: true
> [0:] nativeCharacteristic Notify enabled

However as mentioned the Notification callback is never called so we see no data.

We have put together two samples, one using Plugin.BluetoothLE and one using  Shiny BLE, the Timeout issues do not happen using Plugin.BluetoothLE. Here are some notes on the samples 

BluetoothTest uses Shiny. BluetoothTest2 uses Plugin.BluetoothLE. 

There are 4 buttons in the app

1. Scan
2. Connect
3. Start notifications
4. Disconnect

In our production app, we usually only ever scan once (step 1), storing the discovered UUID for subsequent capture sessionsUsing Plugin.BluetoothLE, we can perform Step 1 then cycle through steps 2, 3 and 4 multiple times. Each time we can read bytes from the device.

This is not the case with Shiny BLE and the call to GetKnownCharacteristic causes the timeout exception.Interestingly, when running the Shiny test app, if after step 4 (disconnect) we scan again (step 1), then connecting to and reading from the device works fine.

[BTLibTests.zip](https://github.com/shinyorg/shiny/files/10696477/BTLibTests.zip)

**Reply by aritchie** [reply]
Thank you for the sample.  I will see what I can do with it in the next week or so

**Reply by aritchie** [reply]
So this issue is being caused by the lack of an old hack that existed in the original plugin that forced the gatt object in android to refresh its internal cache.  A new scan (which internally forces a cache flush) does alleviate this issue.  This hack is "partially" needed for your specific scenario where you actually trigger a disconnect.  With reconnections, the gatt object is never disposed so the issue doesn't seem to occur.

I will add the internal hack in the near future, the issue is that it causes perf issues for other use-cases.  Ideally, you want to rescan when a device is disconnected, so you can get signal strength, device data, etc to show to your user.
